src/inet/mqtt_client.c

```diff
@@ -604,12 +610,40 @@ static int handle_publish(struct mqtt_client_t *self_p,
     }
 
     topic[topic_size] = '\0';
+    qos = ((flags >> 1) & 0x3);
+
+
+    log_object_print(self_p->log_object_p,
+                     LOG_DEBUG,
+                     OSTR("QoS: %d, Flags: %x.\r\n"),
+                     qos,
+                     flags);
+
+    if (qos == 0) {
+        payload_size = (size - topic_size - 2);
+    } else {
+        if (chan_read(self_p->transport.in_p, buf, 2) != 2) { /* read the packet identifier */
+            return (-EIO);
+        }
 
-    if (chan_read(self_p->transport.in_p, buf, 2) != 2) {
-        return (-EIO);
-    }
 
-    payload_size = (size - topic_size - 4);
+        if (qos == 1) {
+            res = write_fixed_header(self_p, MQTT_PUBACK, 0, 2);
+
+        } else { /* qos == 2 */ 
+            res = write_fixed_header(self_p, MQTT_PUBREC, 0, 2);
+        }
```

Comment:
Check that QoS is 2, and add an else that sets res to ``-EPROTO`` since it's a protocol error, right?